### DISCUSSION TITLE
Form validation

### DISCUSSION THREAD

**Discussion by LeaVerou** [Devographics/surveys#120] [State of HTML Features]
Set of HTML attributes and JS APIs to facilitate client-side form validation (`<input pattern>`, `<input required>`, `input. setCustomValidity(msg)` etc)

```html
<input id="zip" pattern="[0-9]{5}" title="5 digits" required />
```

### MDN URL:

https://developer.mozilla.org/en-US/docs/Learn/Forms/Form_validation

### Specification URL:

https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#the-constraint-validation-api

### How stable is this feature?

Stable (Widely supported for several years)

### Type of feature

HTML attribute, JS API, CSS

### What focus area does this feature fall into?

Forms

**Comment by domenic** [comment]
Getting feedback on how this could be improved seems like it would be very valuable to implementers. I can guess developers want more control over styling, wording, etc. but specific details would probably help.

**Reply by LeaVerou** [reply]
That's a good point, if this is in Part 1 of the survey (feature questions), there's not much of that. I wonder if [quick context](https://github.com/Devographics/surveys/discussions/163) could help with that, especially if it can have customized responses for certain questions (e.g. here one of the answers could be "Insufficient styling flexibility", which people could optionally edit to add more details)
